ref: split metadata and add full metadata as default to detector

niermann999 opened the pull request:
For traccc a full detector type should be used that can hold any type of shape or grid that is available in detray. I split off the ```full_metadata``` that is meant to do this from the utils directory and added it to core as a default argument for the detector, so that the comprehensive detector type can be buil as ```detray::detector<>``` .
I added the debug shape types that are only used in detray, so we might want to remove them again from the core type if we continue to use the dedicated telescope type in detray. Currently, I also did not add all grid types that we may get as input from ACTS (all possible binning and boundary type combinations). Once they come up, we need to add them at that point.
This also adds a test of reading in the toy detector from json into the detector type with the full metadata, as I would expect that to be what will happen in traccc in the end?

beomki-yeo commented on core/include/detray/core/detector_metadata.hpp:
```diff
@@ -0,0 +1,250 @@
+
+/// mask types
+/// @TODO: Need to duplicate for pixel/strip measurement dimensions?
+using rectangle = mask<rectangle2D<>, nav_link>;
```
Can we have those type definitions in metadata class or is it still better

niermann999 replied:
This is in the metadata file, or what to you mean exactly?

beomki-yeo replied:
I was meaning that we should avoid the global declarations

niermann999 replied:
Ah, Ok, got it. Yes, you are right

niermann999 commented:
The full metadata distinguishes between cylinders and the simpler portal cylinders, so in order to make them more compatible, I build the toy detector beampipe now as a full cylinder (the intersector will calculate two solutions). This also makes more sense when a track tries to propagate through the beampipe from the outside in and needs to exit it on the other side again.

niermann999 commented:
> The full metadata distinguishes between cylinders and the simpler portal cylinders, so in order to make them more compatible, I build the toy detector beampipe now as a full cylinder (the intersector will calculate two solutions). This also makes more sense when a track tries to propagate through the beampipe from the outside in and needs to exit it on the other side again.

Or not, since the navigation breaks due to #451 I could have known...

beomki-yeo commented on core/include/detray/core/detector_metadata.hpp:
```diff
@@ -0,0 +1,252 @@
+        e_single1 = 8,
+        e_single2 = 9,
+        e_single3 = 10,
+        e_unbounded_rectangle2 = 11,
```
I am not sure if we need to include single (which I don't know how to use) and unbound surface here, but OK

niermann999 commented:
Rebased on grid PR and added the grid types + irregular binning grid types to the full detector metadata. Also harmonized the grid ids between the toy detector and the full detector

niermann999 commented:
Since this got already approved before, let's get this in
